### DISCUSSION TITLE
dropping multirooted SAML message before any processing

### DISCUSSION THREAD

**Discussion by srd90** [node-saml/node-saml#191] [Ideas]
About this approach c1f275c289c01921e58f5c70ce0fdbc5287e5fbe

Is there any scenario where `node-saml` should continue to process incoming message if it has more than one root?

If `node-saml` is configured with `wantAuthnResponseSigned = false` (note: `3.x` series of `passport-saml` [is effetively having this value `false`](https://github.com/node-saml/passport-saml/blob/7b1f23241ebe4d86d4c09c221027bfb9ac54cf98/src/node-saml/saml.ts#L762-L779)) following code block would send potentially harmful XML message "downstream":
https://github.com/node-saml/node-saml/blob/c1f275c289c01921e58f5c70ce0fdbc5287e5fbe/src/saml.ts#L692-L704

I.e. would it be better to move "multiple roots check" from its current location to this line:
https://github.com/node-saml/node-saml/blob/c1f275c289c01921e58f5c70ce0fdbc5287e5fbe/src/saml.ts#L676

and stop processing message immediately if it has more than one root (because it is not well-formed XML).

Related comment from owasp pages:
https://cheatsheetseries.owasp.org/cheatsheets/SAML_Security_Cheat_Sheet.html#validate-signatures
>...
>* Always perform schema validation on the XML document prior to using it for any security-­related purposes:
>   * Always use local, trusted copies of schemas for validation.
>   * Never allow automatic download of schemas from third party locations.
>   * If possible, inspect schemas and perform schema hardening, to disable possible wildcard ­type or relaxed processing statements.
>
>...

Unfortunately javascript ecosystem doesn't seem to have pure-javascript validating XML parser (or parser which would check whether XML document is well-formed see e.g. https://github.com/jindw/xmldom/issues/150 which applies also to [@xmldom/xmldom](https://github.com/xmldom/xmldom)).

sidenote: `samlify` project has taken pluggable validator approach:
https://github.com/tngan/samlify/blob/0950407d57bb1c5a93985a1d9b1bd2fd9182d1d5/README.md#installation

**Comment by cjbarth** [comment]
I've created #195 to address your observation. I put the multiple root check after the check for proper encoding as that allows the error messages to be related to the problem more directly.

**Reply by srd90** [reply]
I found out that @frumioj has also written comments/observations related to multiroot situation handling to here https://github.com/node-saml/passport-saml/commit/8b7e3f5a91c8e5ac7e890a0c90bc7491ce33155e#commitcomment-87082162

This message's purpose is to link those comments to this discussion for easier discoverability of all related comments (seems that  @cjbarth has already noticed those comments).

I suggest that any additional comments would be written to PR #195
